File: ruby/version_metadata_TEST.rb
#!/usr/bin/ruby
# vim: set sw=4 sts=4 et tw=80 :

#
# Copyright (c) 2006 Ciaran McCreesh <ciaranm@ciaranm.org>
#
# This file is part of the Paludis package manager. Paludis is free software;
# you can redistribute it and/or modify it under the terms of the GNU General
# Public License version 2, as published by the Free Software Foundation.
#
# Paludis is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
# details.
#
# You should have received a copy of the GNU General Public License along with
# this program; if not, write to the Free Software Foundation, Inc., 59 Temple
# Place, Suite 330, Boston, MA  02111-1307  USA
#

require 'test/unit'
require 'Paludis'

Paludis::Log.instance.log_level = Paludis::LogLevel::Warning

class Paludis
    class TestCase_VersionMetadata < Test::Unit::TestCase
        def test_no_create
            assert_raise NoMethodError do
                p = VersionMetadata.new
            end
        end

        def env
            unless @env
                @env = NoConfigEnvironment.new("version_metadata_TEST_dir/testrepo/")
            end
            @env
        end

        def vmd version
            env.package_database.fetch_repository("testrepo").version_metadata("foo/bar", version)
        end

        def test_license
            assert_kind_of DepAtom, vmd("1.0").license
        end

        def test_interfaces
            assert vmd("1.0").get_ebuild_interface
            assert ! vmd("1.0").get_ebin_interface
            assert ! vmd("1.0").get_cran_interface
            assert ! vmd("1.0").get_virtual_interface
        end

        def test_members
            assert_equal "Test package", vmd("1.0").description
            assert_equal "http://paludis.berlios.de/", vmd("1.0").homepage
            assert_equal "0", vmd("1.0").slot
            assert_equal "0", vmd("1.0").eapi
            assert_equal "GPL-2", vmd("1.0").license_string
        end

        def test_ebuild_members
            assert_equal "", vmd("1.0").provide_string
            assert_equal "http://example.com/bar-1.0.tar.bz2", vmd("1.0").src_uri
            assert_equal "monkey", vmd("1.0").restrict_string
            assert_equal "test", vmd("1.0").keywords.gsub(%r/\s/, "")
            assert_equal "", vmd("1.0").iuse.gsub(%r/\s/, "")
        end

        def test_deps
            assert_kind_of AllDepAtom, vmd("1.0").build_depend
            assert_kind_of AllDepAtom, vmd("1.0").run_depend
            assert_kind_of AllDepAtom, vmd("1.0").post_depend

            assert_equal 1, vmd("1.0").build_depend.to_a.length
            assert vmd("1.0").run_depend.to_a.empty?
            assert vmd("1.0").post_depend.to_a.empty?

            assert_equal "foo/bar", vmd("1.0").build_depend_string.gsub(/\s/, "")
            assert_equal "", vmd("1.0").run_depend_string.gsub(/\s/, "")
            assert_equal "", vmd("1.0").post_depend_string.gsub(/\s/, "")
        end
    end
end

File: python/version_metadata_TEST.py
#!/usr/bin/env python
# vim: set fileencoding=utf-8 sw=4 sts=4 et :

#
# Copyright (c) 2007 Piotr Jaroszyński <peper@gentoo.org>
#
# This file is part of the Paludis package manager. Paludis is free software;
# you can redistribute it and/or modify it under the terms of the GNU General
# Public License version 2, as published by the Free Software Foundation.
#
# Paludis is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
# details.
#
# You should have received a copy of the GNU General Public License along with
# this program; if not, write to the Free Software Foundation, Inc., 59 Temple
# Place, Suite 330, Boston, MA  02111-1307  USA
#

import os

repo = os.path.join(os.getcwd(), "version_metadata_TEST_dir/testrepo")
irepo = os.path.join(os.getcwd(), "version_metadata_TEST_dir/installed")

from paludis import *
import unittest

Log.instance.log_level = LogLevel.WARNING;

class TestCase_VersionMetadata(unittest.TestCase):
    e = NoConfigEnvironment(repo)
    evdb = NoConfigEnvironment(irepo)

    def vmd(self, version):
        return self.e.package_database.fetch_repository("testrepo").version_metadata("foo/bar", version)

    def vmd_vdb(self, package, version):
        return self.evdb.package_database.fetch_repository("installed").version_metadata(package, version)

    def test_01_get(self):
        self.vmd("1.0")

    def test_02_create_error(self):
        self.assertRaises(Exception, VersionMetadata)

    def test_03_data_members(self):
        vmd = self.vmd("1.0")

        self.assertEquals(str(vmd.slot), "0")
        self.assertEquals(str(iter(vmd.homepage).next()), "http://paludis.pioto.org/")
        self.assertEquals(vmd.description, "Test package")
#        self.assertEquals(vmd.eapi, "0")

    def test_04_ebuild_interface(self):
        ei = self.vmd("1.0").ebuild_interface
        ei2 = self.vmd_vdb("cat-one/pkg-one", "1").ebuild_interface

        self.assert_(isinstance(ei, VersionMetadataEbuildInterface))
        self.assert_(isinstance(ei2, VersionMetadataEbuildInterface))

        self.assertEquals(len(list(ei.provide)), 0)
        self.assertEquals(str(iter(ei.src_uri).next()), "http://example.com/bar-1.0.tar.bz2")
        self.assertEquals(str(iter(ei.restrictions).next()), "monkey")
        self.assertEquals(str(iter(ei.keywords).next()), "test")
        self.assertEquals(len(list(ei.eclass_keywords)), 0)
        self.assertEquals(len(list(ei.iuse)), 0)
        self.assertEquals(len(list(ei.inherited)), 0)

    def test_05_ebin_interface_TODO(self):
        pass

    def test_06_cran_interface_TODO(self):
        pass

    def test_07_deps_interface(self):
        di = self.vmd("1.0").deps_interface
        di2 = self.vmd_vdb("cat-one/pkg-one", "1").deps_interface

        self.assert_(isinstance(di, VersionMetadataDepsInterface))
        self.assert_(isinstance(di2, VersionMetadataDepsInterface))

        self.assert_(isinstance(di.build_depend, AllDepSpec))
        self.assert_(isinstance(di.run_depend, AllDepSpec))
        self.assert_(isinstance(di.suggested_depend, AllDepSpec))
        self.assert_(isinstance(di.post_depend, AllDepSpec))

        self.assertEquals(str(iter(di.build_depend).next()), "foo/bar")
        self.assertEquals(len(list(di.run_depend)), 0)
        self.assertEquals(len(list(di.post_depend)), 0)
        self.assertEquals(len(list(di.suggested_depend)), 0)


    def test_08_origins_interface(self):
        oi = self.vmd_vdb("cat-one/pkg-one", "1").origins_interface

        self.assert_(isinstance(oi, VersionMetadataOriginsInterface))

        self.assertEquals(str(oi.source), "cat-one/pkg-one-1::origin_test")
        self.assertEquals(oi.binary, None)

    def test_09_virutal_interface_TODO(self):
        pass

    def test_10_license_interface(self):
        li = self.vmd("1.0").license_interface
        li2 = self.vmd_vdb("cat-one/pkg-one", "1").license_interface

        self.assert_(isinstance(li, VersionMetadataLicenseInterface))
        self.assert_(isinstance(li2, VersionMetadataLicenseInterface))

        self.assertEquals(str(iter(li.license).next()), "GPL-2")
        self.assertEquals(len(list(li2.license)), 0)

if __name__ == "__main__":
    unittest.main()

